funnel/models/mailer.py

```diff
@@ -0,0 +1,455 @@
+
+    # Customised template for this recipient
+    subject: Mapped[Optional[str]] = sa.orm.mapped_column(
+        sa.Unicode(250), nullable=True
+    )
+    template: Mapped[Optional[str]] = sa.orm.mapped_column(
+        sa.UnicodeText, nullable=True, deferred=True
+    )
+
+    # Rendered version of user's template, for archival
+    rendered_text: Mapped[Optional[str]] = sa.orm.mapped_column(
+        sa.UnicodeText, nullable=True, deferred=True
+    )
+    rendered_html: Mapped[Optional[str]] = sa.orm.mapped_column(
+        sa.UnicodeText, nullable=True, deferred=True
+    )
+
+    # Draft of the mailer template that the custom template is linked to (for updating
+    # before finalising)
+    draft_id: Mapped[Optional[int]] = sa.orm.mapped_column(
+        sa.ForeignKey('mailer_draft.id')
+    )
+    draft: Mapped[Optional[MailerDraft]] = relationship(MailerDraft)
+
+    __table_args__ = (sa.UniqueConstraint('mailer_id', 'url_id'),)
+
+    def __repr__(self) -> str:
+        return f'<MailerRecipient {self.fullname} {self.email} of {self.mailer!r}>'
+
+    @property
+    def fullname(self) -> Optional[str]:
+        """Recipient's fullname, constructed from first and last names if required."""
+        if self._fullname:
+            return self._fullname
+        if self._firstname:
+            if self._lastname:
+                # FIXME: Cultural assumption of <first> <space> <last> name.
+                return f"{self._firstname} {self._lastname}"
+            return self._firstname
+        if self._lastname:
+            return self._lastname
+        return None
+
+    @fullname.setter
+    def fullname(self, value: Optional[str]) -> None:
+        self._fullname = value
+
+    @property
+    def firstname(self) -> Optional[str]:
+        if self._firstname:
+            return self._firstname
+        if self._fullname:
+            return NAMESPLIT_RE.split(self._fullname)[0]
+        return None
+
+    @firstname.setter
+    def firstname(self, value: Optional[str]) -> None:
+        self._firstname = value
+
+    @property
+    def lastname(self) -> Optional[str]:
+        if self._lastname:
+            return self._lastname
+        if self._fullname:
+            return NAMESPLIT_RE.split(self._fullname)[-1]
+        return None
+
+    @lastname.setter
+    def lastname(self, value: Optional[str]) -> None:
+        self._lastname = value
+
+    @property
+    def nickname(self) -> Optional[str]:
+        return self._nickname or self.firstname
+
+    @nickname.setter
+    def nickname(self, value: Optional[str]) -> None:
+        self._nickname = value
+
+    @property
+    def email(self) -> str:
+        return self._email
+
+    @email.setter
+    def email(self, value: str) -> None:
+        self._email = value.lower()
+        self.md5sum = md5sum(value)
+
+    @property
+    def revision_id(self) -> Optional[int]:
+        return self.draft.revision_id if self.draft else None
+
+    def is_latest_draft(self) -> bool:
```

Comment:
Maybe `is_using_latest_draft` and as hybrid property to find all that need updating.